### PR TITLE
PYTHON-1925 Add libmongocrypt cypto hooks

### PR DESCRIPTION
**Pull request by ShaneHarvey**
Jira: https://jira.mongodb.org/browse/PYTHON-1925
Patch: https://evergreen.mongodb.com/version/5d4cb6062a60ed12e5dd77f6

### CODE REVIEW

File: bindings/python/pymongocrypt/binary.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        """Wraps a mongocrypt_binary_t."""
+        self.bin = binary
+        if binary == ffi.NULL:
+            raise MongoCryptError(
+                "unable to create new mongocrypt_binary object")
+
+    def _close(self):
+        """Cleanup resources."""
+        if self.bin is None:
+            return
```
Comment (behackett) [inline bindings/python/pymongocrypt/binary.py]: I kinda hate returns like this. I think putting everything under "if self.bin is not None" reads better.

File: bindings/python/pymongocrypt/binary.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    ffi.memmove(buf, data, len(data))
+
+
+class _MongoCryptBinary(object):
+    __slots__ = ("bin",)
+
+    def __init__(self, binary):
+        """Wraps a mongocrypt_binary_t."""
+        self.bin = binary
+        if binary == ffi.NULL:
```
Comment (behackett) [inline bindings/python/pymongocrypt/binary.py]: Why not do this before setting the instance variable?

File: bindings/python/pymongocrypt/binary.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+
+class MongoCryptBinaryIn(_MongoCryptBinary):
+    __slots__ = ("__copy",)
+
+    def __init__(self, data):
+        """Creates a mongocrypt_binary_t from binary data."""
+        # mongocrypt_binary_t does not own the data it is passed so we need to
+        # create a copy to keep the data alive.
+        self.__copy = ffi.new("uint8_t[]", data)
```
Comment (behackett) [inline bindings/python/pymongocrypt/binary.py]: Does uint8_t work everywhere? For example, Python 2.7 on windows? You might have to use char[].

File: bindings/python/pymongocrypt/binary.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        super(MongoCryptBinaryIn, self).__init__(
+            lib.mongocrypt_binary_new_from_data(self.__copy, len(data)))
+
+    def _close(self):
+        """Cleanup resources."""
+        super(MongoCryptBinaryIn, self)._close()
+        if self.__copy is None:
+            return
```
Comment (behackett) [inline bindings/python/pymongocrypt/binary.py]: Same thing here about bare returns.

File: bindings/python/pymongocrypt/crypto.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+from pymongocrypt.binding import ffi, lib
+from pymongocrypt.compat import str_to_bytes
+
+
+def _callback_error_handler(exception, exc_value, tb):
+    """Set the mongocrypt_status_t on error."""
+    # From cffi docs: "First check if traceback is not None (it is None e.g.
+    # if the whole function ran successfully but there was an error converting
+    # the value returned: this occurs after the call)."
+    if tb is None:
```
Comment (behackett) [inline bindings/python/pymongocrypt/crypto.py]: I love that the cffi docs agree with me about "is not None". :-)

It's also weird that you return None here, but False later.

File: bindings/python/pymongocrypt/crypto.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+@ffi.callback(
+    "bool(void *, mongocrypt_binary_t *, mongocrypt_binary_t *,"
+    "     mongocrypt_binary_t *, mongocrypt_binary_t *, uint32_t *,"
+    "     mongocrypt_status_t *)",
+    onerror=_callback_error_handler)
+def aes_256_cbc_encrypt(ctx, key, iv, input, output, bytes_written, status):
+    backend = default_backend()
+    # TODO: Java uses noPadding but CBC says padding is required:
+    # https://cryptography.io/en/latest/hazmat/primitives/symmetric-encryption/#cryptography.hazmat.primitives.ciphers.modes.CBC
```
Comment (ShaneHarvey) [inline bindings/python/pymongocrypt/crypto.py]: @jyemin This comment is referencing the Java bindings:
https://github.com/mongodb/libmongocrypt/blob/5a2ff2d72c255b0f291891880b5163a8f18ab486/bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/MongoCryptImpl.java#L105-L108

Is that a bug in the Java bindings?

**Comment by ShaneHarvey** [comment]
Updated passing patch: https://evergreen.mongodb.com/version/5d4cbcc10ae6064513aaa022

File: bindings/python/pymongocrypt/crypto.py (hunk 5 above)
Comment (ShaneHarvey) [reply bindings/python/pymongocrypt/crypto.py]: Fixed.

File: bindings/python/pymongocrypt/crypto.py (hunk 6 above)
Comment (ShaneHarvey) [reply bindings/python/pymongocrypt/crypto.py]: Jeff says:
> Its not a bug.  The library does the padding.

**Comment by ShaneHarvey** [comment]
Updated passing patch: https://evergreen.mongodb.com/version/5d4cd8be1e2d172b4fe8fb0b

File: bindings/python/pymongocrypt/crypto.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+from cryptography.hazmat.primitives.hashes import Hash, SHA256, SHA512
+from cryptography.hazmat.primitives.hmac import HMAC
+from cryptography.hazmat.backends import default_backend
+
+from pymongocrypt.binary import _to_bytes, _write_bytes
+from pymongocrypt.binding import ffi, lib
+from pymongocrypt.compat import str_to_bytes
+
+
+def _callback_error_handler(exception, exc_value, tb):
```
Comment (ShaneHarvey) [inline bindings/python/pymongocrypt/crypto.py]: See the cffi docs `onerror` docs for why we set the status via `tb.tb_frame.f_locals['status']`: 
> Note the following hack: in onerror, you can access the original callback arguments as follows. First check if traceback is not None (it is None e.g. if the whole function ran successfully but there was an error converting the value returned: this occurs after the call). If traceback is not None, then traceback.tb_frame is the frame of the outermost function, i.e. directly the frame of the function decorated with @ffi.def_extern(). So you can get the value of argname in that frame by reading traceback.tb_frame.f_locals['argname'].

https://cffi.readthedocs.io/en/latest/using.html?highlight=onerror#extern-python-reference

File: bindings/python/setup.py [hunk 8]
```diff
@@ -19,7 +19,7 @@
     description="Python bindings for libmongocrypt",
     long_description=LONG_DESCRIPTION,
     packages=find_packages(),
-    install_requires=["cffi>=1.12.0,<2"],
+    install_requires=["cffi>=1.12.0,<2", "cryptography>=2.7,<3"],
```
Comment (ShaneHarvey) [inline bindings/python/setup.py]: cryptography 2.7 is the latest release. I haven't attempted to use earlier releases yet but we may be able to do `"cryptography>=2,<3"` instead. That would provide greater compatibility for users who already have cryptography installed.

Comment (behackett) [reply bindings/python/setup.py]: Do some spot testing, back to the earliest 2.x version to ensure they provide what we need.

File: bindings/python/pymongocrypt/binary.py (hunk 3 above)
Comment (behackett) [reply bindings/python/pymongocrypt/binary.py]: Fingers crossed. :-)

File: bindings/python/pymongocrypt/crypto.py (hunk 6 above)
Comment (behackett) [reply bindings/python/pymongocrypt/crypto.py]: Specifically, libmongocrypt is doing the padding. We (or cryptography) can't do further padding on top of that. You should add a comment in our code making the situation clear. I'd hate to have a security researcher opening a security issue over a non-issue.

File: bindings/python/pymongocrypt/binary.py (hunk 3 above)
Comment (ShaneHarvey) [reply bindings/python/pymongocrypt/binary.py]: It seems to work so far on windows + Python 2.7 :)

File: bindings/python/pymongocrypt/crypto.py (hunk 6 above)
Comment (ShaneHarvey) [reply bindings/python/pymongocrypt/crypto.py]: Done.

File: bindings/python/setup.py (hunk 8 above)
Comment (ShaneHarvey) [reply bindings/python/setup.py]: I'll do this in https://jira.mongodb.org/browse/PYTHON-1953.

**Pull request merged**